README.md

# pwrusb

A simple command-line interface for controlling pwrusb.com power strip outlets.

After multiple attempts at a more flexible (but more complicated) interface, decided that it was better to back out to the simplest command-line interface that re-establishes USB connectivity with each call, versus trying to maintain USB connectivity and fall prey to occasional USB dropouts from the pwrusb strip.

## Installation notes

To avoid copyright issues, several files are not included in this distribution.  Makefile will attempt to download and install these files automatically.  Or, they can be downloaded from the Mac software distribution from the [pwrusb website](http://www.pwrusb.com/downloads.html).  These files are:
    
    PwrUSBImp.h
    PwrUsbCmd.cpp
    libpowerusb.dylib

A version of libusb-1.0 will need to be installed.  One straightforward way to get this is via [homebrew](http://brew.sh/) with:

    brew install libusb
            
## Typical Installation

From command line:
    
    make
    make install
    
This will install the executable (`pwrusb`) in `/usr/local/bin/` and the necessary library (`libpowerusb.dylib`) in `/usr/local/lib/'.

## Usage

Example usage:

*Get status:*

    pwrusb status

returns, e.g. (where 0 means Off & 1 means On for outlets 1 2 3):

    Outlet states on bank 0:
        0  0  0

*Set all 3 outlets individually:*

    pwrusb set 0 1 0

where 0 is Off & 1 (or anything other than 0) is On for outlets 1 2 3

*Set all outlets to on/off together:*

    pwrusb setall 0

where 0 is Off & 1 (or anything other than 0) is On

*Set one outlet on/off:*

    pwrusb setone 3 0
    
where first number is outlet number (1 2 or 3) and second is 0 for Off or 1 for On (or anything other than 0)

*Read the total current draw in milliamp:*

    pwrusb current
    
Output looks like:

    Total current (milliamp) on bank 0: 40

Note that current draw will almost never be 0, even if outlets are off.

Note that outlets respond & are controllable even if overall powerstrip red power switch is off.




        
## Additional Usage Notes

Lighted power switch on pwrusb.com strips controls power to outlets, but does not power on/off the USB controller within the strip.  When the physical switch is off, strip should report a low (but not zero) current draw and will report what state the outlets will be when the physical switch is flipped back on, i.e. can report True even when no power is coming out of an outlet.

# History

This package was assembled in mid-2016 from earlier more complicated versions by hroe.  See `CHANGES.txt` for change log.

CHANGES.txt
~~~~~~~~~~~~~~~~~
pwrusb Change Log
~~~~~~~~~~~~~~~~~

------------------
v0.1.0, 2016-09-23
------------------

- Initial working version released publicly

pwrusb-test.cpp
#include <stdio.h>
#include <unistd.h>
#include "_PwrUSBCmd.cpp"

// The purpose of this file is to a quick test case of command line interface

extern int get_number_of_strips_attached()
{
    int i, MaxUnits, connected;
    int ioStates[7];
	char firmware[32];

	memset(ioStates, 0, sizeof(int)*7);					// triggger state of 3 outputs and 4 inputs
    strcpy(firmware, "");
    loadSharedLibrary();
    if ((MaxUnits=(*InitPowerUSB_Address)(&i, firmware)) > 0)		// Initialize the PowerUSB
        connected = (*CheckStatusPowerUSB_Address)();
    return MaxUnits;
}

extern int get_total_current_milliamps(int bank)
{
    int r1, current, i, MaxUnits, connected;
    int ioStates[7];
	char firmware[32];

	memset(ioStates, 0, sizeof(int)*7);					// triggger state of 3 outputs and 4 inputs
    strcpy(firmware, "");
    loadSharedLibrary();
    if ((MaxUnits=(*InitPowerUSB_Address)(&i, firmware)) > 0)		// Initialize the PowerUSB
        connected = (*CheckStatusPowerUSB_Address)();
    // TODO: sometime when I have multiple pwrusb strips attached, test `bank`
    (*SetCurrentPowerUSB_Address)(bank);
    r1 = (*ReadCurrentPowerUSB_Address)(&current);						// Present current consumption reading in milli amps
    return current;
}

int main(int argc, char* argv[])
{
    printf("Yippee yi yo.\n");
    printf("Number of strips attached: %d\n", get_number_of_strips_attached());
    printf("Total current milliamps on bank 0: %d\n", get_total_current_milliamps(0));
    return 0;
}


extern void get_outlet_states(int bank, int *outlet1, int *outlet2, int *outlet3)
{
	int connected=0;
    int in1, in2, in3, i, MaxUnits;
    int ioStates[7];
	char firmware[32];

	memset(ioStates, 0, sizeof(int)*7);					// triggger state of 3 outputs and 4 inputs
    strcpy(firmware, "");
    loadSharedLibrary();

    if ((MaxUnits=(*InitPowerUSB_Address)(&i, firmware)) > 0)		// Initialize the PowerUSB
        connected = (*CheckStatusPowerUSB_Address)();

    // TODO: sometime when I have multiple pwrusb strips attached, test `bank`
    (*SetCurrentPowerUSB_Address)(bank);
    
//     See:
//     http://smile.amazon.com/forum/-/Tx3S42X36ZWYTGH/ref=ask_dp_dpmw_al_hza?asin=B008AZW9P6
// 
//         I am using my PwrUSB connected to a MacMini. They don't support Mac, but
//         they were nice enough to send me the C src code for the programming library.
//         I discovered a bug related ReadPortStatePowerUSB().
// 
//         The read function implementation does not match the documentation. According
//         to the code, the 3 return parameters must go in with values greater than 0
//         otherwise the ports are not actually read. And the returned status is
//         actually the on/off state of the last one read.
// HR: so, therefore, initialize in1,in2,in3 before requesting.  I still haven't figured out
//     what the last comment about 'last one read' means.

    in1 = 999;
    in2 = 999;
    in3 = 999;

    (*ReadPortStatePowerUSB_Address)(&in1, &in2, &in3);

    *outlet1 = in1;
    *outlet2 = in2;
    *outlet3 = in3;
    return;
}


extern void set_outlet_states(int bank, int outlet1, int outlet2, int outlet3)
{
	int connected=0;
    int i, MaxUnits;
    int ioStates[7];
	char firmware[32];

	memset(ioStates, 0, sizeof(int)*7);					// triggger state of 3 outputs and 4 inputs
    strcpy(firmware, "");
    loadSharedLibrary();

    if ((MaxUnits=(*InitPowerUSB_Address)(&i, firmware)) > 0)		// Initialize the PowerUSB
        connected = (*CheckStatusPowerUSB_Address)();
    // TODO: sometime when I have multiple pwrusb strips attached, test `bank`
    (*SetCurrentPowerUSB_Address)(bank);
    (*SetPortPowerUSB_Address)(outlet1, outlet2, outlet3);
    return;
}


extern int get_single_outlet_state(int bank, int outlet_number)
{
    int outlet1, outlet2, outlet3;
    get_outlet_states(bank, &outlet1, &outlet2, &outlet3);
    if (outlet_number == 1){return outlet1;}
    else if (outlet_number == 2){return outlet2;}
    else if (outlet_number == 3){return outlet3;}
    else{return -1;}
}


extern void set_single_outlet_state(int bank, int outlet_number, int state)
{
    int outlet1, outlet2, outlet3, state_checked;
    if (state)
        state_checked = 1;
    else
        state_checked = 0;
    get_outlet_states(bank, &outlet1, &outlet2, &outlet3);
    if (outlet_number == 1){outlet1 = state_checked;}
    else if (outlet_number == 2){outlet2 = state_checked;}
    else if (outlet_number == 3){outlet3 = state_checked;}
    set_outlet_states(bank, outlet1, outlet2, outlet3);
    return;
}


pwrusb-cmdline.cpp
#include <stdio.h>
#include <unistd.h>
#include "_PwrUSBCmd.cpp"

// The purpose of this file is to a quick test case of command line interface

extern int get_number_of_strips_attached()
{
    int i, MaxUnits, connected;
    int ioStates[7];
	char firmware[32];

	memset(ioStates, 0, sizeof(int)*7);					// triggger state of 3 outputs and 4 inputs
    strcpy(firmware, "");
    loadSharedLibrary();
    if ((MaxUnits=(*InitPowerUSB_Address)(&i, firmware)) > 0)		// Initialize the PowerUSB
        connected = (*CheckStatusPowerUSB_Address)();
    return MaxUnits;
}

extern int get_total_current_milliamps(int bank)
{
    int r1, current, i, MaxUnits, connected;
    int ioStates[7];
	char firmware[32];

	memset(ioStates, 0, sizeof(int)*7);					// triggger state of 3 outputs and 4 inputs
    strcpy(firmware, "");
    loadSharedLibrary();
    if ((MaxUnits=(*InitPowerUSB_Address)(&i, firmware)) > 0)		// Initialize the PowerUSB
        connected = (*CheckStatusPowerUSB_Address)();
    // TODO: sometime when I have multiple pwrusb strips attached, test `bank`
    (*SetCurrentPowerUSB_Address)(bank);
    r1 = (*ReadCurrentPowerUSB_Address)(&current);						// Present current consumption reading in milli amps
    return current;
}

extern void get_outlet_states(int bank, int *outlet1, int *outlet2, int *outlet3)
{
	int connected=0;
    int in1, in2, in3, i, MaxUnits;
    int ioStates[7];
	char firmware[32];

	memset(ioStates, 0, sizeof(int)*7);					// triggger state of 3 outputs and 4 inputs
    strcpy(firmware, "");
    loadSharedLibrary();

    if ((MaxUnits=(*InitPowerUSB_Address)(&i, firmware)) > 0)		// Initialize the PowerUSB
        connected = (*CheckStatusPowerUSB_Address)();

    // TODO: sometime when I have multiple pwrusb strips attached, test `bank`
    (*SetCurrentPowerUSB_Address)(bank);
    
//     See:
//     http://smile.amazon.com/forum/-/Tx3S42X36ZWYTGH/ref=ask_dp_dpmw_al_hza?asin=B008AZW9P6
// 
//         I am using my PwrUSB connected to a MacMini. They don't support Mac, but
//         they were nice enough to send me the C src code for the programming library.
//         I discovered a bug related ReadPortStatePowerUSB().
// 
//         The read function implementation does not match the documentation. According
//         to the code, the 3 return parameters must go in with values greater than 0
//         otherwise the ports are not actually read. And the returned status is
//         actually the on/off state of the last one read.
// HR: so, therefore, initialize in1,in2,in3 before requesting.  I still haven't figured out
//     what the last comment about 'last one read' means.

    in1 = 999;
    in2 = 999;
    in3 = 999;

    (*ReadPortStatePowerUSB_Address)(&in1, &in2, &in3);

    *outlet1 = in1;
    *outlet2 = in2;
    *outlet3 = in3;
    return;
}

extern void set_outlet_states(int bank, int outlet1, int outlet2, int outlet3)
{
	int connected=0;
    int i, MaxUnits;
    int ioStates[7];
	char firmware[32];

	memset(ioStates, 0, sizeof(int)*7);					// triggger state of 3 outputs and 4 inputs
    strcpy(firmware, "");
    loadSharedLibrary();

    if ((MaxUnits=(*InitPowerUSB_Address)(&i, firmware)) > 0)		// Initialize the PowerUSB
        connected = (*CheckStatusPowerUSB_Address)();
    // TODO: sometime when I have multiple pwrusb strips attached, test `bank`
    (*SetCurrentPowerUSB_Address)(bank);
    (*SetPortPowerUSB_Address)(outlet1, outlet2, outlet3);
    return;
}

extern void set_single_outlet_state(int bank, int outlet_number, int state)
{
    int outlet1, outlet2, outlet3, new_state;
    if (state)
        new_state = 1;
    else
        new_state = 0;
    get_outlet_states(bank, &outlet1, &outlet2, &outlet3);
    if (outlet_number == 1){outlet1 = new_state;}
    else if (outlet_number == 2){outlet2 = new_state;}
    else if (outlet_number == 3){outlet3 = new_state;}
    set_outlet_states(bank, outlet1, outlet2, outlet3);
    return;
}

void demo()
{
    int outlet1, outlet2, outlet3;
    printf("Yippee yi yo.\n");
    printf("Number of strips attached: %d\n", get_number_of_strips_attached());
    printf("Total current milliamps on bank 0: %d\n", get_total_current_milliamps(0));
    get_outlet_states(0, &outlet1, &outlet2, &outlet3);
    printf("Outlet states on bank 0:  %d  %d  %d\n", outlet1, outlet2, outlet3);
    sleep(1);
    outlet1 = 0;
    outlet2 = 1; 
    outlet3 = 0;
    printf("Setting outlets on bank 0 to:  %d  %d  %d\n", outlet1, outlet2, outlet3);
    set_outlet_states(0, outlet1, outlet2, outlet3);
    sleep(1);
    printf("Outlet states on bank 0:  %d  %d  %d\n", outlet1, outlet2, outlet3);
    sleep(1);
    outlet1 = 1;
    outlet2 = 0; 
    outlet3 = 1;
    printf("Setting outlets on bank 0 to:  %d  %d  %d\n", outlet1, outlet2, outlet3);
    set_outlet_states(0, outlet1, outlet2, outlet3);
    sleep(1);
    printf("Outlet states on bank 0:  %d  %d  %d\n", outlet1, outlet2, outlet3);
}

void status()
{
    int outlet1, outlet2, outlet3;
    int bank=0;
    get_outlet_states(bank, &outlet1, &outlet2, &outlet3);
    printf("Outlet states on bank %d:\n\t%d  %d  %d\n", bank, outlet1, outlet2, outlet3);
}

void version()
{
    printf("pwrusb version is 0.1.0\n");
}

void pwrusb_info()
{
    printf("pwrusb usage examples:\n");
    printf("  pwrusb status\n");
    printf("      returns the current outlet states, in format:\n");
    printf("          Outlet states on bank BANK:\n                OUTLET1  OUTLET2  OULTET3\n");
    printf("      where OUTLET state is 0=off, 1=on\n");
    printf("  pwrusb set OUTLET1STATE OUTLET2STATE OUTLET3STATE\n");
    printf("      sets all 3 outlets to individually specified states\n");
    printf("      where OUTLET[1-3]STATE is 0=off, 1=on  (anything except 0 is on)\n");
    printf("  pwrusb setall STATE\n");
    printf("      sets all 3 outlets\n");
    printf("      where STATE is 0=off, 1=on  (anything except 0 is on)\n");
    printf("  pwrusb setone OUTLETNUM STATE\n");
    printf("      where OUTLETNUM is 1, 2, or 3\n");
    printf("      where STATE is 0=off, 1=on  (anything except 0 is on)\n");
    printf("  pwrusb current\n");
    printf("      returns the total current being drawn (milliamp)\n");
    printf("NOTES: \n");
    printf("    - current is not 0, even when all outlets off, or power strip switch is off\n");
    printf("    - outlets respond & are controllable even if overall powerstrip red power switch is off\n");
    printf("----\n");
    version();
}

int main(int argc, char* argv[])
{
    int outlet1, outlet2, outlet3, outletnum;
    int bank=0;
    
    if (argc == 2 && strcmp(argv[1], "version") == 0)
    {
        version();
        return 0;
    }
    
    if (get_number_of_strips_attached() <= 0)
    {
        pwrusb_info();
        printf("--- No pwrusb strips were detected on USB ----\n");
    }
    else
    if (argc == 2 && strcmp(argv[1], "status") == 0)
        status();
    else 
    if (argc == 5 && strcmp(argv[1], "set") == 0)
    {
        if (strcmp(argv[2], "0") == 0)
            outlet1 = 0;
        else
            outlet1 = 1;
        if (strcmp(argv[3], "0") == 0)
            outlet2 = 0;
        else
            outlet2 = 1;
        if (strcmp(argv[4], "0") == 0)
            outlet3 = 0;
        else
            outlet3 = 1;
        set_outlet_states(bank, outlet1, outlet2, outlet3);
    }
    else
    if (argc == 3 && strcmp(argv[1], "setall") == 0)
    {
        if (strcmp(argv[2], "0") == 0)
            outlet1 = 0;
        else
            outlet1 = 1;
        outlet2 = outlet1;
        outlet3 = outlet1;
        set_outlet_states(bank, outlet1, outlet2, outlet3);
    }
    else    
    if (argc == 4 && strcmp(argv[1], "setone") == 0)
    {
        if ((strcmp(argv[2], "1") != 0) && (strcmp(argv[2], "2") != 0) && (strcmp(argv[2], "3") != 0))
        {
            printf("ERROR: pwrusb setone: unrecognized outlet number, must be 1, 2, or 3\n");
            return -1;
        }
        else
        {
            outletnum = atoi(argv[2]);
            if (strcmp(argv[3], "0") == 0)
                outlet1 = 0;
            else
                outlet1 = 1;
            set_single_outlet_state(bank, outletnum, outlet1);
        }
    }
    else
    if (argc == 2 && strcmp(argv[1], "current") == 0)
    {
        printf("Total current (milliamp) on bank %d: %d\n", bank, get_total_current_milliamps(bank));
    }
    else
        pwrusb_info();
//     demo();    
    return 0;
}







